core/src/main/java/org/infinispan/container/DefaultDataContainer.java

```diff
@@ -100,30 +113,60 @@ protected DefaultDataContainer(int concurrencyLevel, long thresholdSize,
             throw new IllegalArgumentException("No such eviction thread policy " + strategy);
       }
 
-      Eviction eviction;
-      switch (strategy) {
-         case FIFO:
-         case UNORDERED:
-         case LRU:
-            eviction = Eviction.LRU;
+      Caffeine<K, InternalCacheEntry<K, V>> caffeine = caffeineBuilder();
+
+      switch (thresholdPolicy) {
+         case MEMORY:
+            CacheEntrySizeCalculator<K, V> calc = new CacheEntrySizeCalculator<>(new WrappedByteArraySizeCalculator<>(
+                  new MarshalledValueEntrySizeCalculator()));
+            caffeine.weigher((k, v) -> (int) calc.calculateSize(k, v)).maximumWeight(thresholdSize);
             break;
-         case LIRS:
-            eviction = Eviction.LIRS;
-            if (thresholdPolicy == EvictionType.MEMORY) {
-               throw new IllegalArgumentException("Memory based approximation eviction cannot be used with LIRS!");
-            }
+         case COUNT:
+            caffeine.maximumSize(thresholdSize);
             break;
          default:
-            throw new IllegalArgumentException("No such eviction strategy " + strategy);
-      }
-      EntrySizeCalculator<K, InternalCacheEntry<K, V>> sizeCalculator =
-            thresholdPolicy == EvictionType.MEMORY ? new CacheEntrySizeCalculator<>(
-                    new MarshalledValueEntrySizeCalculator()) : null;
+            throw new UnsupportedOperationException("Policy not supported: " + thresholdPolicy);
+      }
+      entries = new CaffeineConcurrentMap<>(applyListener(caffeine, evictionListener).build());
+   }
+
+   private Caffeine<K, InternalCacheEntry<K, V>> applyListener(Caffeine<K, InternalCacheEntry<K, V>> caffeine, DefaultEvictionListener listener) {
+      return caffeine.executor(new WithinThreadExecutor()).removalListener((k, v, c) -> {
```

Comment:
There is an issue I believe with Passivation when an entry is evicted before it has been moved to the back end store.  If we use a different thread there is a timing issue where the object is neither in memory or in the store.  I think I might be able to use a writer only since that is synchronous by default iirc, but I believe that also has an issue.

I can try running the tests without it, but I am not confident it will behave as we need.